Source/santad/EventProviders/SNTDeviceManager.mm

```diff
@@ -0,0 +1,288 @@
+        LOGE(@"Sleeping for 30s before restarting.");
+        sleep(30);
+        exit(ret);
+      default:
+        LOGE(@"Unable to create es client: %d. Sleeping for a minute.", ret);
+        sleep(60);
+        continue;
+    }
+  }
+}
+
+- (void)listenES API_AVAILABLE(macos(10.15)) {
+  while (!self.client)
+    usleep(100000);  // 100ms
+
+  es_event_type_t events[] = {
+    ES_EVENT_TYPE_AUTH_MOUNT,
+  };
+
+  es_return_t sret = es_subscribe(self.client, events, sizeof(events) / sizeof(es_event_type_t));
+  if (sret != ES_RETURN_SUCCESS)
+    LOGE(@"SNTDeviceManager: unable to subscribe to auth mount events: %d", sret);
+}
+
+- (void)listenDA {
+  DARegisterDiskAppearedCallback(_diskArbSession, NULL, diskAppearedCallback,
+                                 (__bridge void *)self);
+  DARegisterDiskDescriptionChangedCallback(_diskArbSession, NULL, NULL,
+                                           diskDescriptionChangedCallback, (__bridge void *)self);
+  DARegisterDiskDisappearedCallback(_diskArbSession, NULL, diskDisappearedCallback,
+                                    (__bridge void *)self);
+}
+
+- (void)listen {
+  [self listenDA];
+  if (@available(macos 10.15, *)) [self listenES];
+  self.subscribed = YES;
+}
+
+- (void)handleAuthMount:(const es_message_t *)m
+             withClient:(es_client_t *)c API_AVAILABLE(macos(10.15)) {
+  if (!self.blockUSBMount) {
+    es_respond_auth_result(self.client, m, ES_AUTH_RESULT_ALLOW, false);
+    return;
+  }
+
+  long mountMode = m->event.mount.statfs->f_flags;
+  pid_t pid = audit_token_to_pid(m->process->audit_token);
+  LOGI(@"SNTDeviceManager: mount syscall arriving from path: %s, pid: %d, fflags: %lu",
+       m->process->executable->path.data, pid, mountMode);
+
+  DADiskRef disk =
+    DADiskCreateFromBSDName(NULL, self.diskArbSession, m->event.mount.statfs->f_mntfromname);
+  CFAutorelease(disk);
+
+  // TODO(tnek): Log all of the other attributes available in diskInfo into a structured log format.
+  NSDictionary *diskInfo = CFBridgingRelease(DADiskCopyDescription(disk));
+  BOOL isRemovable = [diskInfo[(__bridge NSString *)kDADiskDescriptionMediaRemovableKey] boolValue];
+  BOOL isUSB = [diskInfo[@"DADeviceProtocol"] isEqualTo:@"USB"];
+
+  if (!isRemovable || !isUSB) {
+    es_respond_auth_result(self.client, m, ES_AUTH_RESULT_ALLOW, false);
+    return;
+  }
+
+  if (self.remountArgs != nil && [self.remountArgs count] > 0) {
+    long remountOpts = mountArgsToMask(self.remountArgs);
+    if (mountMode & remountOpts) {
+      es_respond_auth_result(self.client, m, ES_AUTH_RESULT_ALLOW, false);
+      return;
+    }
+
+    long newMode = mountMode | remountOpts;
+    LOGI(@"SNTDeviceManager: remounting device '%s'->'%s', flags (%lu) -> (%lu)",
+         m->event.mount.statfs->f_mntfromname, m->event.mount.statfs->f_mntonname, mountMode,
+         newMode);
+    [self remount:disk mountMode:newMode];
+  }
+  es_respond_auth_result(self.client, m, ES_AUTH_RESULT_DENY, false);
```

Comment:
Move this before the if on line 213.

Not sure we need to worry about the timeout if we do this.

Reply:
Done